proposals/hasCrossSiteAncestor.md

```diff
@@ -0,0 +1,122 @@
+
+```
+// Cookie details object with partitionKey containing hasCrossSiteAncestor value of true.
+"details": {
+  "partitionKey": {
+    "topLevelSite": "https://example.com",
+    "hasCrossSiteAncestor": true
+  }
+}
+```
+
+### New Permissions
+No new permissions are required.
+
+### Manifest File Changes
+No new manifest fields are required.
+
+## Security and Privacy
+Privacy: The hasCrossSiteAncestor boolean value does not reveal private information.
+
+Security: The hasCrossSiteAncestor boolean value does not impact security.
+
+### Exposed Sensitive Data
+No sensitive data is associated with the cross-site ancestor chain bit.
+
+### Abuse Mitigations
+A cookie may only have no cross-site ancestor when the topLevelSite in the partitionKey and the URL with which the cookie is associated with are first-party to each other. To prevent the creation of cookies that violate this, the set method will return an error if the URL and the topLevelSite are not first-party for cookies that are set with no-cross site ancestor.
+
+There are situations that can occur (such as an update to the public suffix list) which can change the if the topLevelSite and the URL are first-party to each other. To accomidate this possibility, the value for `hasCrossSiteAncestor` for the get(), getAll() and remove() methods will not be restricted. This will allow for web extensions to migrate or remove cookies that have become invalid after they have been set.
+
+### Additional Security Considerations
+None
+
+## Alternatives
+The hasCrossSiteAncestor value could be a value that is not exposed to extensions at all. Browsers that choose to include a cross-site ancestor chain bit in their partitioned cookies, could calculate the expected value of the cross-site ancestor chain bit from the URL associated with the cookie and the topLevelSite in the partitionKey. However, this could cause extensions to be unable to correctly set or get cookies (in an A1->B->A2 situation) as the browser may calculate the incorrect value for hasCrossSiteAncestor since it would not be explicitly provided by the extension.
+
+### Existing Workarounds
+To access cookies with the same `topLevelSite` but different `hasCrossSiteAncestor` values (A1->B->A2 context), developers can remove the `Partitioned` attribute from the cookie and use the Storage Access API in the context of web pages.
+
+### Open Web API
+The APIs being expanded to include the hasCrossSiteAncestor boolean are specific to extensions.
+
+## Implementation Notes
+
+### Populating hasCrossSiteAncestor when not provided
+When no value has been provided for `hasCrossSiteAncestor`, if the `domain` associated with the `cookie` is same-site to the value of the `topLevelSite`, the `hasCrossSiteAncestor` value will be set to false otherwise the value will be set to true.
```

Comment:
It makes sense to me that including a list of valid combinations is a good idea for our own alignment and to help developers understand the API. I've listed the combinations below for chrome. 

Valid representations of a partitionKey with hasCrossSiteAncestor included for Chrome
- Partitioned Cookie
  -  `{"topLevelSite": "https://example.com", "hasCrossSiteAncestor": false}`
      - Partitioned cookie in a same-site context, such as a partitioned cookie set on the top-level site.
  -  `{"topLevelSite": "https://example.com", "hasCrossSiteAncestor": true}`
      - Partitioned cookie in a cross-site context, such as a partitioned cookie set in a cross-site embed.
- Unpartitioned Cookie partitionKey
  - `{"topLevelSite": "", "hasCrossSiteAncestor": false}`
  - These representations will have the same behavior as the one above
    - `{}`
    - `{"topLevelSite": "", "hasCrossSiteAncestor": true}`
    - `{"hasCrossSiteAncestor": true}`
    - `{"hasCrossSiteAncestor": false}`

Reply:
I see an interesting difference - the combination that I classified as invalid before appears to be valid in Chrome, as the first bullet point of your list. I suppose that if Firefox were to implement `Partitioned`, that the "invalid" case would suddenly become valid. Firefox Nightly already supports the Partitioned attribute, and I can confirm that I also see a topLevelSite in this case - TIL.

Support for top-level Partitioned means that a top-level document (and same-site child frames, potentially nested) always has access to at least TWO sets of cookies: Unpartitioned, and partitioned with `hasCrossSiteAncestor:false`. I can indeed confirm that I see two cookies when I run the following:

```javascript
document.cookie = "a=one; SameSite=None; Secure; Partitioned";
document.cookie = "a=two";
document.cookie // "a=two; a=one" in Firefox, "a=one; b=two" in Chrome.
```

Reply:
Since non-URL-serialization `topLevelSites` throw, I would expect `{"topLevelSite": "", "hasCrossSiteAncestor": false}`, `{"topLevelSite": "", "hasCrossSiteAncestor": true}`, `{"hasCrossSiteAncestor": true}`, and `{"hasCrossSiteAncestor": false}` to all throw an exception.

Reply:
Invalid keys should result in a throw:
- `{"hasCrossSiteAncestor": false}` and `{"hasCrossSiteAncestor": true}` are invalid because the `hasCrossSiteAncestor` value is only valid within the context of the full key and there is no backward compatibility that needs to be supported.
- `{"topLevelSite": "", "hasCrossSiteAncestor": true}` is invalid because an unpartitioned cookie can't have a cross-site ancestor.

`{"topLevelSite": "", "hasCrossSiteAncestor": false}` would be considered a valid key because it reflects the key described as a valid key for an unpartitioned cookie in the background information section of the document.

Reply:
Cool! That just leaves `{"topLevelSite": "", "hasCrossSiteAncestor": false}` as a gap.

To me, `{"topLevelSite": "", "hasCrossSiteAncestor": false}` is the same as `{"hasCrossSiteAncestor": false}`- neither set the full key. And[ I think the background section identifies both as valid.](https://github.com/w3c/webextensions/pull/581/files#diff-6dddb2726fe8306191f3e612a738e1b31e904af6344489bdaf93c5dbd411db80R14). 

Is there a positive reason to have this? It seems like a non-empty partitonKey object being "unpartitioned" is more confusing than helpful.

Reply:
I'll update the language to make it clear that a key with a `hasCrossSiteAncestor` value and no `topLevelSite` is always invalid. 

Both the Firefox and chromium implementations support the empty topLevelSite partitionKey (`{"topLevelSite": ""}`)  for this API. Since it is already supported, it made sense to me to allow developers to also pass the valid value for the `hasCrossSiteAncestor` for that key.

Reply:
Ah, I misunderstood the existing behavior and thought `{"topLevelSite": ""}` would throw in exiting implementations. If that's the case, this makes sense to me!

Reply:
Added a table showing the valid keys with a description of the type of cookie they are associated with.

Reply:
@Rob--W, @bvandersloot-mozilla, gave this some more thought and removed the empty key from the table of valid partitionKeys. Since it's not used to indicate a specific partitionKey but instead used by the existing implementations of `cookies.remove()` and `cookies.getAll()` to indicate that the api should be accessing unpartitioned and partitioned cookies.

Added the information for the empty partitionKey to the implementation notes instead.

Reply:
In Firefox, get/set/remove with an empty object for partitionKey is equivalent to passing `null` and means unpartitioned cookies. The remove() method removes only one cookie.

getAll() differs from the above; the object passed to getAll is treated like a filter. The use of a partitionKey property signals that the extension is aware of partitioning and passing `{}` as partitionKey causes it to return partitioned and unpartitioned cookies.

Reply:
I'm fine with adding it back to the table and have it indicate an unpartitioned cookie with the note that for `cookies.getAll()` it returns partitioned and unpartitioned cookies.